firecracker-microvm/firecracker
# jailer: implement seccomp filtering logic

# Changes made
- added BPF macro-like functions that generate the seccomp filters bytecode for the scenarios relevant to Firecracker - validate the architecture, allow a set of syscalls and deny the others by having the kernel emit SIGSYS

## What's missing
- integration with the jailer: will add once the jailer binary becomes available
- integration tests: will add once the jailer binary becomes available
- `SIGSYS` handler: will be addressed in a separate PR

# Testing done
# Unit & integration tests
```bash
$  sudo env PATH=$PATH python3 -m pytest # all passed
```
| File | Coverage |
|------|----------|
| seccomp.rs | 72.2% |
| total | 74.1% |

## Manual tests

Create a mini jailer:

```bash
$ mkdir src/bin/demo_jailer && cat <<EOF  src/bin/demo_jailer/main.rs
 extern crate sys_util;
 use std::env::args;
 use std::os::unix::process::CommandExt;
 use std::process::exit;
 use std::process::{Command, Stdio};
 fn main() {
 let args: Vec<String = args().collect();
 let exec_file = &args[1];
 match sys_util::setup_seccomp() {
     Ok(_) = (),
     Err(_) = {
         exit(1);
 }}
 Command::new(exec_file).stdin(Stdio::inherit()).stdout(Stdio::inherit()).stderr(Stdio::inherit()).exec();
 exit(1);
 }
 EOF
$ sed -i 's/vmm = { path = "vmm" }/vmm = { path = "vmm" }\nsys_util = { path = "sys_util" }/' Cargo.toml 
$ cargo build
```

Jail firecracker and verify that it works correctly

```bash
$ strace -ff -o strace.out target/x86_64-unknown-linux-musl/debug/demo_jailer target/x86_64-unknown-linux-musl/debug/firecracker --api-sock /tmp/firecracker.socket
```
```bash
$ curl --unix-socket /tmp/firecracker.socket -i  \
      -X PUT "http://localhost/boot-source" \
      -H "accept: application/json" \
      -H "Content-Type: application/json" \
      -d "{ 
            \"boot_source_id\": \"alinux_kernel\",
            \"source_type\": \"LocalImage\", 
            \"local_image\": 
                 { 
                     \"kernel_image_path\": \"${kernel_path}\" 
                 }
         }"
HTTP/1.1 201 Created
Content-Length: 0
Date: Thu, 07 Jun 2018 07:59:29 GMT

$ curl --unix-socket /tmp/firecracker.socket -i \
      -X PUT "http://localhost/machine-config" \
      -H "accept: application/json" \
      -H "Content-Type: application/json" \
      -d "{ \"vcpu_count\": 4, \"mem_size_mib\": 256}"
HTTP/1.1 204 No Content
Date: Thu, 07 Jun 2018 07:59:33 GMT

$ curl --unix-socket /tmp/firecracker.socket -i \
      -X PUT "http://localhost/drives/root" \
      -H "accept: application/json" \
      -H "Content-Type: application/json" \
      -d "{ 
             \"drive_id\": \"root\",
             \"path_on_host\": \"${rootfs_path}\", 
             \"is_root_device\": true, 
             \"permissions\": \"rw\", 
             \"state\": \"Attached\"
          }"
HTTP/1.1 201 Created
Content-Length: 0
Date: Thu, 07 Jun 2018 07:59:38 GMT

$ curl --unix-socket /tmp/firecracker.socket -i \
      -X PUT "http://localhost/drives/read_only_drive" \
      -H "accept: application/json" \
      -H "Content-Type: application/json" \
      -d "{ 
             \"drive_id\": \"read_only_drive\",
             \"path_on_host\": \"${ro_drive_path}\", 
             \"is_root_device\": false, 
             \"permissions\": \"ro\", 
             \"state\": \"Attached\"
          }"
HTTP/1.1 201 Created
Content-Length: 0
Date: Thu, 07 Jun 2018 07:59:42 GMT

$ curl --unix-socket /tmp/firecracker.socket -i \
      -X PUT "http://localhost/network-interfaces/1" \
      -H "accept: application/json" \
      -H "Content-Type: application/json" \
      -d "{ 
             \"iface_id\": \"1\", 
             \"host_dev_name\": \"vmtap33\", 
             \"guest_mac\": \"06:00:00:00:00:01\", 
             \"state\": \"Attached\" 
         }"
HTTP/1.1 201 Created
Content-Length: 0
Date: Thu, 07 Jun 2018 08:00:42 GMT

$ curl --unix-socket /tmp/firecracker.socket -i \
      -X PUT "http://localhost/actions/start" \
      -H  "accept: application/json" \
      -H  "Content-Type: application/json" \
      -d "{  
             \"action_id\": \"start\",  
             \"action_type\": \"InstanceStart\"
          }"
HTTP/1.1 201 Created
Content-Length: 0
Date: Thu, 07 Jun 2018 08:00:46 GMT

$ curl --unix-socket /tmp/firecracker.socket -i \
      -X GET "http://localhost/actions/start" \
      -H "accept: application/json"
HTTP/1.1 200 OK
Content-Type: application/json
Transfer-Encoding: chunked
Date: Thu, 07 Jun 2018 08:00:52 GMT

{"action_id":"start","action_type":"InstanceStart","timestamp":0}

$ top -H -n 1 -p `pidof firecracker
  PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND
  16989 aghecen   20   0  295944 116560 115676 S 13.3  0.7   0:04.16 fc_vcpu0
  16508 aghecen   20   0  295944 116560 115676 S  0.0  0.7   0:00.02 firecracker
  16509 aghecen   20   0  295944 116560 115676 S  0.0  0.7   0:00.83 firecracker
  16990 aghecen   20   0  295944 116560 115676 S  0.0  0.7   0:00.00 fc_vcpu1
  16991 aghecen   20   0  295944 116560 115676 S  0.0  0.7   0:00.00 fc_vcpu2
  16992 aghecen   20   0  295944 116560 115676 S  0.0  0.7   0:00.00 fc_vcpu3
$ killall firecracker
```

Inspect the syscalls that made it
```bash
$ for i in strace.*; do cat $i >> strace.out; done
$ cat strace.out | cut -d\( -f1 | sort | uniq
accept
arch_prctl
bind
brk
clone
close
dup
epoll_create1
epoll_ctl
epoll_pwait
eventfd2
execve
fcntl
fstat
futex
getrandom
ioctl
+++ killed by SIGTERM +++
listen
lseek
mmap
mprotect
munmap
open
pipe
prctl
read
readlink
readv
rt_sigaction
rt_sigprocmask
sched_getaffinity
set_tid_address
sigaltstack
--- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=18808, si_uid=628827462} ---
--- SIGWINCH {si_signo=SIGWINCH, si_code=SI_KERNEL} ---
socket
stat
write
writev
```

Jail a program that tries to call something else

```bash
$ strace -o strace_ls.out target/x86_64-unknown-linux-musl/debug/demo_jailer /bin/ls
Bad system call
$ tail -5 strace_ls.out 
execve("/bin/ls", ["/bin/ls"], 0x7fffa3008230 /* 65 vars */) = 0
brk(NULL)                               = 0xdda000
access("/etc/ld.so.nohwcap", F_OK)      = 21
--- SIGSYS {si_signo=SIGSYS, si_code=SYS_SECCOMP, si_call_addr=0x7fecc39873c7, si_syscall=__NR_access, si_arch=AUDIT_ARCH_X86_64} ---
+++ killed by SIGSYS +++
```
